design/data_only_restore.md

```diff
@@ -0,0 +1,141 @@
+
+``` go
+type RestoreSpec struct {
+    ......
+	// DataOnly specify whether to only restore volume data and related k8s resources
+	// without any other k8s resources.
+	DataOnly *bool `json:"dataOnly"`
+    ......
+}
+```
+
+### Restore CLI change
+The `--data-only` parameter is added to the `velero restore create` CLI.
+Both `--data-only` and `--data-only=true` mean enabling the data-only restore feature.
+Both `--data-only=false` and not specifying the parameter mean disabling the data-only restore feature.
+
+The `velero restore describe` CLI should display the backed-up PVC and restored PVC mapping result.
+The `velero restore describe` CLI should also display the `--data-only` parameter value.
+
+### Resource Filtering
+If the data-only feature is enabled (`restore.Spec.DataOnly` is set to `true`) in the restore, the Velero server will set the `restore.Spec.IncludedResources` to include the data-only related resources. The resource list is `PersistentVolumeClaims`, `PersistentVolumes`, `VolumeSnapshots`, `VolumeSnapshotContents`, `VolumeSnapshotClasses`, and `Secrets`.
+
+When the data-only feature is enabled, the restore must not set the `IncludedResources` and the `ExcludedResources` parameters.
+
+The restore provided filters only apply to the referenced backup including the `PersistentVolumeClaims` resource. The user can use the filters(including `IncludedNamespaces`, `ExcludedNamespaces`, `LabelSelector` and `OrLabelSelectors`) to select which volume data to restore.
+
+### Generate data-only restore's PVC mapping result
+`dataOnlyPvcMap` is added to the `restoreContext` to record the data-only restore generated PVC mapping between the PVC in the backup.
+``` go
+type restoreContext struct {
+    ......
+    dataOnlyPvcMap map[types.NamespacedName]types.NamespaceName
+}
+```
+
+The function `restoreItem` should rename the PVC and record the mapping into `restoreContext.dataOnlyPvcMap` on the PVC resource.
+
+After the restore completes, the PVC mapping result should be stored in the restore result structure, and persist into the `restore-<restore-name>-result.json.gz` metadata.
+``` go
+type Result struct {
+	// Velero is a slice of messages related to the operation of Velero
+	// itself (for example, messages related to connecting to the
+	// cloud, reading a backup file, etc.)
+	Velero []string `json:"velero,omitempty"`
+
+	// Cluster is a slice of messages related to backup or restore of
+	// cluster-scoped resources.
+	Cluster []string `json:"cluster,omitempty"`
+
+	// Namespaces is a map of namespace name to slice of messages
+	// related to backup or restore namespace-scoped resources.
+	Namespaces map[string][]string `json:"namespaces,omitempty"`
+
+    // DataOnlyPvcMap is a map of backed-up PVC name and the data-only
+    // restored PVC name.
+    DataOnlyPvcMap map[types.NamespacedName]types.NamespaceName `json:"dataOnlyPvcMap,omitempty"`
```

Comment:
It should be possible to add labelSelector to all resources that user expect to be swapping to restored PVC name, and having velero do the swapping on labeled (core k8s only for now) resources in the namespace.

```
kubectl label deployment deployName velero.io/sync-restored-pvc-name=<original-pvc-in-backup-name>
```

Any PVC restored using this method with a new name will have annotation `velero.io/restored-pvc-original-name=<original name>` that velero can use as reference for subsequent backup/restore pvc name syncing.

Reply:
As we go with the in-place way, the name-mapping logic is unnecessary.

Reply:
right.. for FSB with deleting PVC prior to restoring to the same name.